### PR TITLE
Sperf & Cperf Benchmarks

### PR DESCRIPTION
**Pull request by trexfeathers**
### Description

* Migrates the SPerf and CPerf benchmarks for the UK Met Office NG-VAT project.
* Refactors some benchmark modules to account for these additions.
* Adds the `@on_demand_benchmark` decorator, since S/CPerf benchmarks are inappropriate for inclusion by default.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

**Comment by trexfeathers** [comment]
_Think_ I've solved the git nightmare 😫. @pp-mo if you're running through commit-by-commit then 51f8a8f25b7890786626b2c20e393b73a9cb5da0 is the 'beginning'.

File: benchmarks/benchmarks/sperf/combine_regions.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        self, n_cubesphere, imaginary_data=True, create_result_cube=False
+    ):
+        # In this case only, do *not* create the result cube.
+        # That is the operation we want to test.
+        super().setup(n_cubesphere, imaginary_data, create_result_cube)
+
+    def time_create_combined_cube(self, n_cubesphere):
+        self.recombine()
+
+    def track_addedmem_create_combined_cube(self, n_cubesphere):
```
Comment (pp-mo) [inline benchmarks/benchmarks/sperf/combine_regions.py:188]: Can we not use our magic decorator approach on these too? #magicdecorator
It seems not to be used in this file.
I think we are also anyway no longer labelling these track results with a 'Mb' unit -- though we were, where this code originally came from in the 'metrics' repo

File: benchmarks/benchmarks/sperf/combine_regions.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+@on_demand_benchmark
+class ComputeRealData(Mixin):
+    """
+    Time+memory costs of computing combined-regions data.
+    """
+
+    def time_compute_data(self, n_cubesphere):
+        _ = self.recombined_cube.data
+
+    def track_addedmem_compute_data(self, n_cubesphere):
```
Comment (pp-mo) [inline benchmarks/benchmarks/sperf/combine_regions.py:203]: #magicdecorator

File: benchmarks/benchmarks/sperf/combine_regions.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    array data, so that input data is not loaded from disk during the save
+    operation.
+
+    """
+
+    def time_save(self, n_cubesphere):
+        # Save to disk, which must compute data + stream it to file.
+        save(self.recombined_cube, "tmp.nc")
+
+    def track_addedmem_save(self, n_cubesphere):
```
Comment (pp-mo) [inline benchmarks/benchmarks/sperf/combine_regions.py:223]: #magicdecorator

File: benchmarks/benchmarks/sperf/combine_regions.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    ):
+        # In this case only, do *not* replace the loaded regions data with
+        # 'imaginary' data, as we want to test file-to-file calculation+save.
+        super().setup(n_cubesphere, imaginary_data, create_result_cube)
+
+    def time_stream_file2file(self, n_cubesphere):
+        # Save to disk, which must compute data + stream it to file.
+        save(self.recombined_cube, "tmp.nc")
+
+    def track_addedmem_stream_file2file(self, n_cubesphere):
```
Comment (pp-mo) [inline benchmarks/benchmarks/sperf/combine_regions.py:253]: #magicdecorator

**Review by pp-mo — CHANGES_REQUESTED** [review]
As noted OWOK I think.
the definition of the 'decorator' in  TrackAddedMemoryAllocation is cool !

I'm also left a bit uncertain as to whether the division into overnight/on-demand is sufficiently clear.  
Will discuss offline ...

File: benchmarks/benchmarks/sperf/combine_regions.py (hunk 1 above)
Comment (trexfeathers) [reply benchmarks/benchmarks/sperf/combine_regions.py:189]: bebf01b157d82bbdcff32b3b3b80179f03865a83

File: benchmarks/benchmarks/sperf/combine_regions.py (hunk 2 above)
Comment (trexfeathers) [reply benchmarks/benchmarks/sperf/combine_regions.py:203]: bebf01b157d82bbdcff32b3b3b80179f03865a83

File: benchmarks/benchmarks/sperf/combine_regions.py (hunk 4 above)
Comment (trexfeathers) [reply benchmarks/benchmarks/sperf/combine_regions.py:249]: bebf01b157d82bbdcff32b3b3b80179f03865a83

**Review by pp-mo — APPROVED** [review]

**Comment by trexfeathers** [comment]
We did it! Thanks so much @pp-mo, I know you've had a lot to do.

**Pull request merged**